src/drivers/uavcan_v1/SubscriptionManager.hpp

```diff
@@ -0,0 +1,130 @@
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file SubscriptionManager.hpp
+ *
+ * Manages the UAVCAN subscriptions
+ *
+ * @author Peter van der Perk <peter.vanderperk@nxp.com>
+ */
+
+#pragma once
+
+
+#include <px4_platform_common/defines.h>
+#include <drivers/drv_hrt.h>
+#include "Subscribers/DynamicPortSubscriber.hpp"
+#include "CanardInterface.hpp"
+
+#include "ServiceClients/GetInfo.hpp"
+#include "ServiceClients/Access.hpp"
+#include "Subscribers/BaseSubscriber.hpp"
+#include "Subscribers/Heartbeat.hpp"
+#include "Subscribers/DS-015/Battery.hpp"
+#include "Subscribers/DS-015/Esc.hpp"
+#include "Subscribers/DS-015/Gnss.hpp"
+#include "Subscribers/legacy/LegacyBatteryInfo.hpp"
+#include "Subscribers/uORB/sensor_gps.hpp"
+
+typedef struct {
+	const char *px4_name;
+	UavcanDynamicPortSubscriber *(*create_sub)(CanardInstance &ins, UavcanParamManager &pmgr) {};
+} UavcanDynSubBinder;
+
+class SubscriptionManager
+{
+public:
+	SubscriptionManager(CanardInstance &ins, UavcanParamManager &pmgr) : _canard_instance(ins), _param_manager(pmgr) {}
+
+	void subscribe();
+	void printInfo();
+	void updateParams();
+
+private:
+	CanardInstance &_canard_instance;
+	UavcanParamManager &_param_manager;
+	UavcanDynamicPortSubscriber *_dynsubscribers {NULL};
+
+	UavcanHeartbeatSubscriber _heartbeat_sub {_canard_instance};
+	// GetInfo response
+	UavcanGetInfoResponse _getinfo_rsp {_canard_instance};
+
+	// Process register requests
+	UavcanAccessResponse  _access_rsp {_canard_instance, _param_manager};
+
+	const UavcanDynSubBinder _uavcan_subs[6] {
+		{
+			"UCAN1_ESC0_PID",
+			[](CanardInstance & ins, UavcanParamManager & pmgr) -> UavcanDynamicPortSubscriber *
+			{
+				return new UavcanEscSubscriber(ins, pmgr, 0);
+			}
+		},
```

Comment:
What ideas have come up around being able to include only select subscriber classes at compile time?  I think we're close to being able to do that.  Perhaps a linked list rather than a static array, with `#ifdef CONFIG_*` around the inclusion and addition of each subscriber into the list?

Reply:
Indeed pretty much adding #ifdef CONFIG_* guards here would do the trick, that's why I migrated all subscribers to this class. Only thing is to deal with the duplicate definitaion of the PX4 parameter in parameters.c and ParamManager.cpp, maybe we could make some template for it.